### PR TITLE
Don't lose cart after logging in

### PR DESCRIPTION
**Pull request by artursmet**
Scenario:
* Start as an anonymous user
* Add something to cart, proceed to checkout
* Log in using first step of the checkout

Failure: You'll be redirected to cart page showing empty cart
Expected result: Your anonymous cart is now your current cart as logged in user

### CODE REVIEW

File: saleor/cart/utils.py [hunk 1]
```diff
@@ -65,20 +65,31 @@ def check_product_availability_and_warn(request, cart):
...
+                return
+            cart = get_anonymous_cart_from_token(
+                token=token, cart_queryset=queryset)
+            if cart is None:
+                return
+            cart.change_user(request.user)
+            carts_to_close = Cart.objects.open().filter(user=request.user)
+            carts_to_close = carts_to_close.exclude(token=token)
+            carts_to_close.update(
+                status=Cart.CANCELED, last_status_change=now())
```
Comment (patrys) [inline saleor/cart/utils.py]: Lines 82-86 are likely candidates for `transaction.atomic`.

File: saleor/cart/utils.py [hunk 2]
```diff
@@ -65,20 +65,31 @@ def check_product_availability_and_warn(request, cart):
...
-    cart.change_user(request.user)
-    carts_to_close = Cart.objects.open().filter(user=request.user)
-    carts_to_close = carts_to_close.exclude(token=token)
-    carts_to_close.update(status=Cart.CANCELED, last_status_change=now())
+    def get_cart(view):
+        @wraps(view)
+        def func(request, *args, **kwargs):
+            token = request.get_signed_cookie(Cart.COOKIE_NAME, default=None)
+            if not token:
+                return
```
Comment (patrys) [inline saleor/cart/utils.py]: What do we return here?

File: saleor/cart/utils.py [hunk 3]
```diff
@@ -65,20 +65,31 @@ def check_product_availability_and_warn(request, cart):
...
+            cart = get_anonymous_cart_from_token(
+                token=token, cart_queryset=queryset)
+            if cart is None:
+                return
+            cart.change_user(request.user)
```
Comment (patrys) [inline saleor/cart/utils.py]: How do we know the user is logged in by this point? We have yet to execute `view(...)`.

**Comment by patrys** [comment]
It seems that even after the fixes the decorator often return `None` without ever calling the underlying view.

**Comment by artursmet** [comment]
@patrys I added more tests to cover this scenario

File: saleor/cart/utils.py [hunk 4]
```diff
@@ -65,20 +77,34 @@ def check_product_availability_and_warn(request, cart):
...
-    cart.change_user(request.user)
-    carts_to_close = Cart.objects.open().filter(user=request.user)
-    carts_to_close = carts_to_close.exclude(token=token)
-    carts_to_close.update(status=Cart.CANCELED, last_status_change=now())
+    def get_cart(view):
+        @wraps(view)
+        def func(request, *args, **kwargs):
+            response = view(request, *args, **kwargs)
+            token = request.get_signed_cookie(Cart.COOKIE_NAME, default=None)
+            if not token or token is not None and not token_is_valid(token):
```
Comment (patrys) [inline saleor/cart/utils.py]: `if not token or not token_is_not_valid(token):`?

Comment (patrys) [inline saleor/cart/utils.py]: What if I am logged in and have a cart assigned but don't have a cookie?

Comment (artursmet) [inline saleor/cart/utils.py]: If you are authenticated and don't have a cookie token, then database cart is taken into account

**Comment by littlehome-eugene** [comment]
login = find_and_assign_anonymous_cart()(allauth_login) 

I think it only covers social login, but not regular login,

Wouldn't it be better to place `find_and_assign_anonymous_cart` in user_logged_in signal?
not sure if you need to hook it up with user_signed_up signal as well.

**Comment by artursmet** [comment]
@littlehome-eugene It covers standard login too, in fact it was tested and designed for regular login. When we use django-allauth, standard login() view from django.contrib.auth is not used

**Pull request merged**